src/features/stories/storiesComponents/SourceBlock.tsx

```diff
@@ -0,0 +1,300 @@
+        )
+      ) {
+        if (chapter <= 2 && newTabId === SideContentType.substVisualizer) {
+          toggleStoriesUsingSubst(true, env);
+        }
+
+        if (prevTabId === SideContentType.substVisualizer) {
+          toggleStoriesUsingSubst(false, env);
+        }
+
+        setSelectedTab(newTabId);
+      }
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  );
+
+  const chapterVariantDisplay = chapter ? styliseSublanguage(chapter, variant) : '';
+
+  const dataVisualizerTab: SideContentTab = {
+    label: 'Data Visualizer',
+    iconName: IconNames.EYE_OPEN,
+    body: <SideContentDataVisualizer />,
+    id: SideContentType.dataVisualizer
+  };
+
+  const envVisualizerTab: SideContentTab = {
+    label: 'Env Visualizer',
+    iconName: IconNames.GLOBE,
+    body: <SideContentEnvVisualizer />,
+    id: SideContentType.envVisualizer
+  };
+
+  const processStepperOutput = (output: InterpreterOutput[]) => {
+    const editorOutput = output[0];
+    if (
+      editorOutput &&
+      editorOutput.type === 'result' &&
+      editorOutput.value instanceof Array &&
+      editorOutput.value[0] === Object(editorOutput.value[0]) &&
+      isStepperOutput(editorOutput.value[0])
+    ) {
+      return editorOutput.value;
+    } else {
+      return [];
+    }
+  };
+
+  const tabs = React.useMemo(() => {
+    const tabs: SideContentTab[] = [];
+
+    // For HTML Chapter, HTML Display tab is added only after code is run
+    if (chapter === Chapter.HTML) {
+      if (output.length > outputIndex && output[outputIndex].type === 'result') {
+        tabs.push({
+          label: 'HTML Display',
+          iconName: IconNames.MODAL,
+          body: (
+            <SideContentHtmlDisplay
+              content={(output[outputIndex] as ResultOutput).value}
+              handleAddHtmlConsoleError={errorMsg =>
+                dispatch(addHtmlConsoleError(errorMsg, env, true))
+              }
+            />
+          ),
+          id: SideContentType.htmlDisplay
+        });
+      }
+      return tabs;
+    }
+
+    // (TEMP) Remove tabs for fullJS until support is integrated
+    if (chapter === Chapter.FULL_JS) {
+      return [...tabs, dataVisualizerTab];
+    }
+
+    if (chapter >= 2) {
+      // Enable Data Visualizer for Source Chapter 2 and above
+      tabs.push(dataVisualizerTab);
+    }
+    if (chapter >= 3 && variant !== Variant.CONCURRENT && variant !== Variant.NON_DET) {
+      // Enable Env Visualizer for Source Chapter 3 and above
+      tabs.push(envVisualizerTab);
+    }
+
+    if (chapter <= 2 && (variant === Variant.DEFAULT || variant === Variant.NATIVE)) {
+      // Enable Subst Visualizer only for default Source 1 & 2
+      tabs.push({
+        label: 'Stepper',
+        iconName: IconNames.FLOW_REVIEW,
+        body: <SideContentSubstVisualizer content={processStepperOutput(output)} />,
+        id: SideContentType.substVisualizer
+      });
+    }
+
+    return tabs;
```

Comment:
awesome that story blocks have access to debugger tools like the playground! would there be future plans to add breakpoint support to story blocks for env viz?

also would be great if there can be a simple demo video to show the different functionalities and how we can use these debugger tools in stories! _right now i can't seem to get the stepper (added a video for ref below) and data viz to work but i have a feeling i'm doing something wrong_

https://user-images.githubusercontent.com/47914880/230731477-d3a9f993-3bb9-4be1-abff-644ce60c5399.mov